Transactions categories

klaidas01 opened the pull request.

klaidas01 committed:
revert unintended changes

klaidas01 committed:
add category object

klaidas01 committed:
add category get and post methods

klaidas01 committed:
add category validation

klaidas01 committed:
add default category list

Martynas-S commented on backend/src/modules/transactions/schemas/category.schema.ts:
```diff
@@ -0,0 +1,7 @@
+export const CategorySchema = new mongoose.Schema({
+  Name: String,
+  Icon: String,
+  _id: String,
```
Sitas man atrodo yra nereikalingas: https://mongoosejs.com/docs/guide.html#_id

Martynas-S commented on backend/src/modules/transactions/dto/transaction.dto.ts:
```diff
@@ -21,7 +21,7 @@ export class TransactionDTO {
   @ApiModelProperty()
   @IsString()
-  readonly Category: string;
+  Category: any;
```
manau reiketu palikti kazkoki type. Jei cia laikomas kategorijos `id` tai gali likti `string`, nes schemoje jis bus `string`. O jei ne `string`, tai ka ten virsuj veikia `@IsString()`? :DD

Andrius-B replied:
man atrodo @klaidas01 planavo embeddint tas kategorijas į transaction dto objektą, bet man kažkaip atrodo, kad mes galim biškį kitaip šitą padaryt, gal daugiau papasakosiu per standup'ą mūsų

Martynas-S commented on backend/src/modules/transactions/service/category.service.ts:
```diff
@@ -0,0 +1,73 @@
+      { _id: '0', Name: 'Transport', Icon: 'PlaceholderIcon0' },
+      { _id: '1', Name: 'Food', Icon: 'PlaceholderIcon1' },
+      { _id: '2', Name: 'Entertainment', Icon: 'PlaceholderIcon2' },
+      { _id: '3', Name: 'Health', Icon: 'PlaceholderIcon3' },
```
Galejo buti tikros ikonos :D

Andrius-B replied:
Tikra tiesa, for reference, būtų gerai jei išrinktum iconus ir surašytum jų pavadinimus čia vietoj place-holderių
For reference, fronte dbr naudojami icon'ai iš [čia](https://materialdesignicons.com) ir [čia](https://material.io/tools/icons)  tai reikėtų tokių pavadinimų kaip `mdi/food-apple` arba `google/info` respectiveley, kad žinotume iš kur paimt front-end'e (pridėk tipo pradžioj iš kur imi)

Martynas-S reviewed (commented):
Sita siaip turetu paziureti @Andrius-B

Andrius-B commented on backend/src/modules/transactions/transactions.controller.ts:
```diff
@@ -34,17 +41,64 @@ export class TransactionsController {
   })
   @ApiResponse({ status: 400, description: 'Bad request.' })
-  @UsePipes(new ValidationPipe(this.logger))
+  @UsePipes(new ValidationPipe(new Logger()))
```
O tai tas injected `logger`'is neveikė kažkodėl?

Andrius-B commented on backend/src/modules/transactions/transactions.controller.ts:
```diff
@@ -23,9 +27,12 @@ import { TransactionQueryService } from './service/transaction-query.service';
     private readonly transactionsService: TransactionsService,
+    private readonly categoryService: CategoryService,
     private readonly queryService: TransactionQueryService,
     private readonly logger: Logger,
```
ar šito čia ne-injectina?

Andrius-B commented on backend/src/modules/transactions/transactions.controller.ts:
```diff
@@ -34,17 +41,64 @@ export class TransactionsController {
     return res.status(HttpStatus.CREATED).send();
   }
 
+  @Post('/categories')
```
šiaip gal sakyčiau, kad šiti turtėtų būt perkelti į failą `categories.controller.ts`

Andrius-B reviewed (changes requested):
būtų labai gerai jei čia iškart parinktum icon'us kategorijom, nors ir nieko nedarys dabar tie icon'ai

klaidas01 committed:
Add category icons

Andrius-B commented on backend/src/modules/transactions/service/category.service.ts:
```diff
@@ -47,15 +47,15 @@ export class CategoryService {
+      { _id: '5', Name: 'Bills', Icon: 'mdi/clipboard-text' },
+      { _id: '6', Name: 'Travel', Icon: 'mdi/wallet-travel' },
+      { _id: '7', Name: 'Clothes', Icon: 'mdi/hanger' },
+      { _id: '8', Name: 'Personal care', Icon: 'mdi/dumbbell' },
```
perrrrfect!

Martynas-S replied:
O kaip del likusiu dalyku kuriuos komentavai? :D

klaidas01 committed:
create categories controller

Martynas-S commented on backend/src/modules/transactions/categories.controller.ts:
```diff
@@ -0,0 +1,70 @@
+    description: 'Category successfully received.',
+  })
+  @ApiResponse({ status: 400, description: 'Bad request.' })
+  @UsePipes(new ValidationPipe(new Logger()))
```
Cia butu geriau, man atrodo, permesti `this.logger`, o ne `new Logger()`

klaidas01 committed:
use the same instance of logger for category service

klaidas01 committed:
use ref to link category and transaction schemas and use populate to load category data

klaidas01 committed:
remove unecessary categories import, use populate to load category data in findById method

Martynas-S commented on backend/src/modules/transactions/transactions.controller.ts:
```diff
@@ -23,9 +27,12 @@ import { TransactionQueryService } from './service/transaction-query.service';
     private readonly logger: Logger,
-  ) {}
+  ) {
+    this.categoryService.loadCategories();
```
As tik dabar supratau, kad tu seedini categories duombazes duomenis `TransactionController`. Tai yra negerai del keleto dalyku pvz.: nereikalingas performance hit ( tu atlieki operacijas kurias reikia atlikti **viena karta** per visa aplikacijos gyvavimo laika kiekviena karta kai transactionControlleris yra sukuriamas), nebegalima buna istrinti default kategoriju, neiseina priskirti atskiru kategoriju atskiriem useriam ir t.t.

Sprendimas kurio tu ieskai sitai problemai yra vadinamas `Database seeding`. Look it up

Martynas-S replied:
Be to sitoj vietoj reiketu apsispresti ar reikia is viso reikia db dabar. Jei darom be db, tai viskas labai paprastai, categoryService tures savyje ta hardocdinta masyva, ir tokie dalykai kaaip findAll categories tiesiog grazins ta masyva. Zymiai paprasciau ir tai yra pusiau reliable variantas,  jeigu nenorim tvarkingai seedint duombazes.

klaidas01 replied:
Bet tada nesigaus, kad nera kaip prideti nauju kategoriju ir kaip editinti bazines? O jeigu seedint db tai cia turetu buti kazkoks visiskai atskiras skriptas?

Martynas-S replied:
Jo, jei be db darom tai negalim editint ir kurt nauju kategoriju. Tai in the long run, butu geriau tureti db bet reikia teisingai padaryti.
O seedinimas tia jo, butu atskiras kazkoks scriptas/toolas, kuri cj reiketu parunnint viena karta pries pirma karta leidziant backa, koks nors `npm run seed`

- https://www.google.lt/search?ei=y1rNXMKLA-qymwXn0pPoCA&q=seeding+mong+odb
- https://www.npmjs.com/package/mongoose-seed
- https://docs.mongodb.com/manual/reference/program/mongoimport/

Cia kazkokie linkai, reikia daugiau pasiskaityt

Martynas-S replied:
Cj butu gerai taip padaryt kaip sitas vyrukas daro:
https://www.youtube.com/watch?v=V30Rpqi6kYE
Visus tuos hardcodintus kur dabar turi kontroleryje taip i seed faila ir tuomet padarai `package.json` scripta `seed: node seeds/categories.js` or something kas paleistu ir viskas bus ciki

klaidas01 committed:
add category seeder

klaidas01 committed:
remove loadCategories method from category service

Martynas-S commented:
Oi ups, as dar norejau pora dalyku pakomentuot ;D

Martynas-S commented on backend/package.json:
```diff
@@ -19,7 +19,8 @@
     "format": "prettier --write \"src/**/*.ts\"",
-    "format:tslint": "tslint --fix --project tsconfig.json; prettier --write \"src/**/*.ts\""
+    "format:tslint": "tslint --fix --project tsconfig.json; prettier --write \"src/**/*.ts\"",
+    "seed": "node src/seeds/categories.js"
```
Ye, sitas yra labai gerai, bet tuomet nereikalingas yra tas libas `mongoose-data-seed` ir visi jo konfiguracijos failai kaip `backend/md-seed-config.js` ir t.t. tai reiketu juos istrinti

Martynas-S commented on backend/src/modules/transactions/models/category.model.js:
```diff
@@ -0,0 +1,10 @@
+var mongoose = require('mongoose');
+var Schema = mongoose.Schema;
+
+var schema = new Schema({
```
Uhmm kodel tu duplikuoji schema kuri jau yra apibrezta `backend/src/modules/transactions/schemas/category.schema.ts` ? Reikia sita trint lauk

Martynas-S commented on backend/src/modules/transactions/schemas/category.schema.ts:
```diff
@@ -0,0 +1,7 @@
+export const CategorySchema = new mongoose.Schema({
+  Name: String,
+  Icon: String,
+  _id: String,
```
Dar sitas neresolvintas ;D

Martynas-S replied:
> Sitas man atrodo yra nereikalingas: https://mongoosejs.com/docs/guide.html#_id

klaidas01 replied:
Kazkodel pasalinus id lauka is schemos jis grazina objektus isvis be id ir niekaip nerandu kodel

Martynas-S commented on backend/src/seeds/categories.js:
```diff
@@ -0,0 +1,73 @@
+var mongoose = require('mongoose');
+var Category = require('../modules/transactions/models/category.model');
```
Idomiausias klausimas cia yra kodel tu pradejai naudoti ES5 sintakse ir CommonJS modules kai visa laika buvo ES6 ir ES6 Modules :D. Ka turiu omeny, tai kodel `var` o ne `const` ir kodel `require` o ne `import`. TSLintas turetu mesti milijona erroru uz tokius dalykus

Martynas-S commented on backend/src/seeds/categories.js:
```diff
@@ -0,0 +1,73 @@
+  categories[i].save(function(err, result) {
+    done++;
+    if (done === categories.length) {
+      exit();
```
Sitas `exit()` turetu buti iskeltas is conditional sakinio ir for loopo, kad visada butu grazus cleanup

klaidas01 replied:
Tame tutoriale, pagal kuri dariau, sake, kad iskelus is loopo gali atsijungti nuo db nesukelus visu objektu

Martynas-S replied:
Ai nu jo, tiesa :+1:

Martynas-S reviewed (changes requested):
Siaip tai bendrai gerai, biski keleta dalyku dar reikia pakeist.
Dar vienas svarbus siaip dalykas kuri turetu daryti sitas PRas tai integruoti naujus category modelius i musu fronta. Jei nori tu gali padaryti, jei nenori tai kas nors kitas arba as galim uzbaigt

klaidas01 committed:
use ES6 syntax for category seeder

klaidas01 committed:
remove mongoose-data-seed

Martynas-S commented on backend/package.json:
```diff
@@ -40,7 +40,6 @@
     "joi": "^14.3.1",
     "mongodb": "^3.1.13",
     "mongoose": "^5.4.14",
-    "mongoose-data-seed": "^1.0.7",
```
Is `package.json` paremovinai, bet `package-lock.json` liko visi change'ai del sito package :D Reiktu ir `package-lock.json` sutaisyt.

Martynas-S replied:
No worries, sutvarkiau uz tave su 4355711635158411620f0923970702974ff79889

Martynas-S committed:
Merge branch 'development' into transactions_categories

Martynas-S committed:
Make `_id` generation automatic in `CategorySchema`

Martynas-S committed:
Remove transaction service population with categories

Because currently the functionality provided by populating transactions
with categories serves no purpose, because there is no use case in the
frontend where each transaction should have the whole of the category
object attached to itself.

Besides it:
- Incurs a performance penalty for each transaction model query by
  searching over the same categories collection multiple times over
- Breaks the Transaction interface
- Is not fully fleshed out, e.g. the TransactionQueryService is missing
  the popoulation of categories

Martynas-S committed:
Change frontend interfaces to account for Category string id

Martynas-S committed:
Remove unused test file TransactionList.test.tsx

Martynas-S committed:
Made the ICategory interface in the frontend contain the icon

Also pretty much rewrote the <TransactionListItemIcon /> component.
The main changes were to clean up the logic and remove the style props
for the component, because since it is not a generic component but
rather a domain specific one (Icon for TransactionListItem, not generic
Icon) it shouldn't accept style props, but it can handle it's one style
transparently, by itself, otherwise it exposes what should be considered
as its private API

Martynas-S committed:
Change category database seed to use the correct format of MDIs

Martynas-S committed:
Change `other` category icon to `mdiBankTransfer`

Martynas-S committed:
Connect categories in the frontend to the API

This patch **breaks backwards compatibility**

Martynas-S commented:
The changes in the PR **are not backwards compatible**. 
Once checking out these changes one should:
- Purge the mongo database of any documents in the `categories` collection (if any): via this command in the mongo shell: `db.categories.remove({})`
- Seed the database with categories by running `npm run seed` in the `/backend` directory
- _Preferably_ purge the database of old transactions referring to nonexisting categories by executing this command in the mongo shell `db.transactions.remove({})`
